File: Day 02/d2.hs
import System.IO

fromStringToTuple :: String -> (Char, Int)
fromStringToTuple str = (head str , getValue str)

getValue :: String -> Int
getValue str = read $ cutDirection str

cutDirection :: String -> String
cutDirection (h:t)
        | h == ' ' = t
        | otherwise = cutDirection t

compute :: [(Char, Int)] -> (Int, Int, Int, Int) -- (posx, depth, depth2, aim)
compute [] = (0, 0, 0, 0)
compute ((d, v):t)
  | d == 'f' = (posx + v, depth, depth2 + (aim*v), aim)
  | d == 'u' = (posx, depth - v, depth2, aim - v)
  | d == 'd' = (posx, depth + v, depth2, aim + v)
  where (posx, depth, depth2, aim) = compute t

solvePart1 :: [String] -> Int
solvePart1 l = posx * depth
        where (posx, depth, _, _) = compute $ fromStringToTuple <$> l

solvePart2 :: [String] -> Int
solvePart2 l = posx * depth
        where (posx, _, depth, _) = compute $ reverse $ fromStringToTuple <$> l

main = do
        input <- lines <$> readFile "input.txt"
        print $ solvePart1 input
        print $ solvePart2 input

File: Day 02/d2.c
#include <stdio.h>

#define MAXBUFFER 20

int main(){
  FILE * inputFile = fopen("./input.txt", "r");
  if(inputFile == NULL){
    printf("File not found");
    return 0;
  }
  char line[MAXBUFFER];
  char direction[MAXBUFFER];
  int value = 0, posx = 0, depth = 0, depth2 = 0, aim = 0;
  while(fgets(line, MAXBUFFER, inputFile) != NULL){
    sscanf(line, "%s %d\n", direction, &value);
    switch(direction[0]){
      case 'f':
          posx += value;
          depth2 += aim*value;
          break;
      case 'u':
          depth -= value;
          aim -= value;
          break;
      case 'd':
          depth += value;
          aim += value;
            break;
    }
  }
  fclose(inputFile);
  printf("%d\n", posx*depth);
  printf("%d\n", posx*depth2);
}
